sdc/datatypes/hpat_pandas_dataframe_functions.py

```diff
@@ -978,14 +1003,52 @@ def _df_getitem_slice_idx_impl(self, idx):
     return func_text, global_vars
 
 
-def gen_df_getitem_slice_idx_impl(self):
-    func_text, global_vars = df_getitem_str_slice_codegen(self)
+def df_getitem_tuple_idx_codegen(self, idx):
+    """
+    Example of generated implementation with provided index:
+        def _df_getitem_tuple_idx_impl(self, idx)
+          res_index = self._index
+          data_1 = get_dataframe_data(self, 1)
+          res_data_1 = pandas.Series(data_1, index=res_index, name="B")
+          data_2 = get_dataframe_data(self, 2)
+          res_data_2 = pandas.Series(data_2, index=res_index, name="C")
+          return pandas.DataFrame({"B": res_data_1, "C": res_data_2}, index=res_index)
+    """
+    func_lines = ['def _df_getitem_tuple_idx_impl(self, idx):']
+    literal_idx = {col.literal_value for col in idx}
+    key_error = any(i not in self.columns for i in literal_idx)
 
-    loc_vars = {}
-    exec(func_text, global_vars, loc_vars)
-    _impl = loc_vars['_df_getitem_slice_idx_impl']
+    if self.columns and not key_error:
+        func_lines += df_getitem_tuple_idx_main_codelines(self, literal_idx)
+    else:
+        # raise KeyError if input DF is empty or idx is invalid
+        func_lines += ['  raise KeyError']
+
+    func_text = '\n'.join(func_lines)
+    global_vars = {'pandas': pandas, 'numpy': numpy,
+                   'get_dataframe_data': get_dataframe_data}
+
+    return func_text, global_vars
+
+
+def gen_df_getitem_impl_generator(codegen, impl_name):
```

Comment:
This looks as a very generic function, that we could use everywhere else, so it might be useful to move it to sdc_typing_utils. So it's responsibility would be to call exec on provided func_code and return back created func object.